pkg/controller/nodekeeper/nodekeeper_controller.go

```diff
@@ -0,0 +1,225 @@
+		// Cluster is not upgrading, no error occurred.
+		return reconcile.Result{}, nil
+	}
+
+	reqLogger.Info("Cluster is upgrading. Proceeding.")
+
+	// Initialise metrics
+	metricsClient, err := r.metricsClientBuilder.NewClient(r.client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	pdbForceDrainTimeout, err = getPDBForceDrainTimeout(r.client, reqLogger)
+	if err != nil {
+		if err == upgradeConfigNotfound {
+			reqLogger.Info(upgradeConfigNotfound.Error())
+			reqLogger.Info("Requeuing")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	reqLogger.Info("Checking for PodDisruptionBudget alerts.")
+
+	found, err := checkPDBAlerts(metricsClient)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !found {
+		log.Info("Found no PDB alerts. No further action")
+		return reconcile.Result{}, nil
+	}
+
+	log.Info(fmt.Sprintf("Found PodDisruptionBudgetAtLimit Alert, Force Drain at: %d minutes", pdbForceDrainTimeout))
+
+	return reconcile.Result{}, nil
+
+}
+
+// isClusterUpgrading determines if a cluster is upgrading by
+// - checking CVO conditions for an upgrading state
+// - checking machineconfigpools for an upgrading state
+func isClusterUpgrading(c client.Client, log logr.Logger) (bool, error) {
+	cV, err := osd_cluster_upgrader.GetClusterVersion(c)
+	if err != nil {
+		log.Info("Failed to retrieve ClusterVersion CR")
+		return false, err
+	}
+
+	var conditionProgressing configv1.ClusterOperatorStatusCondition
+	// Iterate through Conditions until we find progressing
+	for _, c := range cV.Status.Conditions {
+		if c.Type == configv1.OperatorProgressing {
+			conditionProgressing = c
+		}
+	}
+
+	// Check if progressing is true.
+	if conditionProgressing.Status != configv1.ConditionTrue {
+		log.Info("Cluster is not upgrading. No further action.")
+		return false, nil
+	}
+
+	yes, err := areNodesUpgrading(c, "worker", log)
+	if err != nil {
+		// An error occurred, return it.
+		return false, err
+	} else if !yes {
+		// Nodes are stil upgrading, no error.
+		return false, nil
+	}
+
+	// Nodes have completed upgrade, no error.
+	return true, nil
+}
+
+func checkPDBAlerts(mC metrics.Metrics) (bool, error) {
+	pdbAlertName := "PodDisruptionBudgetAtLimit"
+	var pdbAlerts *metrics.AlertResponse
+	pdbAlerts, err := mC.Query(fmt.Sprintf("ALERTS{alertname=\"%s\"}", pdbAlertName))
+
+	if err != nil {
+		return false, err
+	}
+
+	if len(pdbAlerts.Data.Result) == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func getUpgradeConfig(c client.Client, logger logr.Logger) (*upgradev1alpha1.UpgradeConfig, error) {
+	uCList := &upgradev1alpha1.UpgradeConfigList{}
+
+	err := c.List(context.TODO(), uCList)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(uCList.Items) == 0 {
+		return nil, upgradeConfigNotfound
+	}
+
+	return &uCList.Items[0], nil
+}
+
+// areNodesUpgrading checks whether nodes are upgraded
+func areNodesUpgrading(c client.Client, nodeType string, logger logr.Logger) (bool, error) {
```

Comment:
Since this is being used in the osd upgrader too, can you create a pkg and interface for this and use it in both places?
```
lib/
└── machineconfig
    └── mc.go
```

This sound ok @mrbarge ?

Reply:
Definitely supportive of that, reduces the duplicated logic.

Reply:
resolved: 18a3825c33a3c99da7b8c7b5dd26992128533a46